Scheduling of Live Events(#77)

Deep-Chill opened the pull request:
- Early versions of API endpoints/URLs to add and delete LiveEvents.
- Gets LiveEvent objects using their scheduled_at time, for adding/deleting as well as marking attendance.
- Updates Mixin to add a rank property to help correctly add attributes to LiveEvent objects.

Issues:
 - On the URL: `http://127.0.0.1:8000/sequences/steve/seq-cert-live-event/1/` if a user is not logged in, they are able to make POST requests and the timer updates, but if logged in as either steve or a superuser it returns `djaodjin-pages-vue.js:581 Error sending ping: Forbidden` Haven't investigated yet.

Deep-Chill committed:
[Work-in-Progress] Scheduling of Live Events(#77)

- Early versions of API endpoints/URLs to add and delete LiveEvents.
- Gets LiveEvent objects using their scheduled_at time, for adding/deleting as well as marking attendance.
- Updates Mixin to add a rank property to help correctly add attributes to LiveEvent objects.

- Issues:
- - On the URL: `http://127.0.0.1:8000/sequences/steve/seq-cert-live-event/1/` if a user is not logged in, they are able to make POST requests and the timer updates, but if logged in as either steve or a superuser it returns `djaodjin-pages-vue.js:581 Error sending ping: Forbidden` Haven't investigated yet.

Deep-Chill committed:
Allows adding/removing LiveEvents and other functionalities

- Added LiveEvent creation/deletion API Endpoints
- Added ability to view multiple LiveEvents belonging to the same PageElement on the front-end view. Could implement this with an API endpoint instead and call that  for views individual users. Needs logic updating to be functional, ie. dealing with PageElements without liveevents and other edge cases
- Context-specific comments in code with comments and questions. To be removed before final version.

Deep-Chill committed:
[Work-in-Progress] Fixes to LiveEvent

Currently working on:
- fixing issue where EnumeratedProgress is not keeping track of the exact LiveEvent. Current approach is to add an extra field in the EnumeratedProgress model and use that.
- Ensuring the delete method changes the status of the LiveEvent and doesn't delete it
- Correct URL locations

Deep-Chill committed:
Updates LiveEvent APIs, serializers, and models

Added soft delete to LiveEvents, updates to attendee retrieval logic and more.

Notes:
- Need further clarification on URL patterns.
- Not sure if a soft delete is required for LiveEvent attendance.
- A few comments in there, to be removed before merge.

smirolo commented on pages/api/progress.py:
```diff
@@ -201,7 +204,7 @@ def post(self, request, *args, **kwargs):
 
 
-class LiveEventAttendanceAPIView(EnumeratedProgressRetrieveAPIView):
+class LiveEventAttendanceAPIView(EnumeratedProgressRetrieveAPIView, DestroyAPIView):
```
It seems that the inheritance should be `(DestroyModelMixin, EnumeratedProgressRetrieveAPIView)` otherwise we have two base `View` classes.

smirolo commented on pages/api/progress.py:
```diff
@@ -224,12 +227,26 @@ class LiveEventAttendanceAPIView(EnumeratedProgressRetrieveAPIView):
+    def get_object(self):
+        return self.progress
 
     def get_serializer_class(self):
```
The whole function might be removed and `serializer_class = ...` added if necessary.

smirolo commented on pages/api/progress.py:
```diff
@@ -224,12 +227,26 @@ class LiveEventAttendanceAPIView(EnumeratedProgressRetrieveAPIView):
+        progress = self.progress
+        live_event_rank = get_extra(progress, "live_event_rank")
+
+        if live_event_rank != self.kwargs.get(self.event_rank_kwarg):
```
I don't understand this code. Why is there a `live_event_rank` in `progress.extra`? Couldn't the code just be something like get_object_or_404(self.progress.live_events, rank=self.kwargs.get(self.event_rank_kwarg))`?

Deep-Chill replied:
The idea here was to check that we're retrieving the correct EnumeratedProgress object, ie. one whose 'live_event_rank' in the extra field matches the live event rank in the URL, if it exists.

For example, one page element has 2 live events(ranks 1 and 2).
We access the URL to retrieve it, api_mark_attendance(args=page element, live event(rank=1))
If we retrieve the progress for that page element, this logic here returns it only if the enumeratedprogress's extra field contains live event(rank=1), meaning progress exists for that specific live event, and wouldn't return anything for api_mark_attendance(args=page element, live event(rank=2)).

smirolo replied:
Got it. Please add a comment along those lines where the `EnumeratedProgress.extra` field is updated, and make sure if we try to change the value of `EnumeratedProgress.extra['live_event_rank']` after it has been set once, there is some kind of warning/error being raised. Thank you!

smirolo commented on pages/api/progress.py:
```diff
@@ -253,21 +270,77 @@ def post(self, request, *args, **kwargs):
         progress = self.get_object()
         element = progress.step
-        live_event = LiveEvent.objects.filter(element=element.content).first()
+        live_event = get_object_or_404(
```
This is the code I would have expected in previous comment.

smirolo commented on pages/api/progress.py:
```diff
@@ -253,21 +270,77 @@ def post(self, request, *args, **kwargs):
-            progress.viewing_duration <= element.min_viewing_duration):
-            progress.viewing_duration = element.min_viewing_duration
-            progress.save()
+        if live_event.scheduled_at >= datetime_or_now():
```
1. Typically prefer to add a statement `at_time = datetime_or_now()` at the entry point of the method, then use `at_time` everywhere else. This way we only have one 'now' in the whole HTTP request.

smirolo commented on pages/api/progress.py:
```diff
@@ -253,21 +270,77 @@ def post(self, request, *args, **kwargs):
+
+        extra["live_event_rank"] = event_rank
+        progress.extra = json.dumps(extra)
+        progress.save()
```
That seems very clanky to save a "live_event_rank" in the EnumeratedProgress. What happens if I ping pong between marking attendees for the Monday and Wednesday live events for example?

smirolo commented on pages/api/progress.py:
```diff
@@ -253,21 +270,77 @@ def post(self, request, *args, **kwargs):
+        queryset = queryset.filter(id__in=progress_ids)
+        return queryset
+
+    def get(self, request, *args, **kwargs):
```
This method does not seem necessary as `LiveEventAttendeesAPIView` inherits from `ListAPIView`.

smirolo commented on pages/api/sequences.py:
```diff
@@ -401,9 +404,9 @@ def post(self, request, *args, **kwargs):
-        return api_response.Response(
-            serializer.errors,
-            status=status.HTTP_400_BAD_REQUEST)
+        # return api_response.Response(
```
This is odd the method does not return any result. Maybe the code at line 399 (HTTP_201_CREATED) should be placed here.

smirolo commented on pages/api/sequences.py:
```diff
@@ -432,3 +435,47 @@ def perform_destroy(self, instance):
+    def get(self, request, *args, **kwargs):
+        return self.list(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
```
Use the redefinition of `ListCreateAPIView.post` to add API documentation.

smirolo commented on pages/api/sequences.py:
```diff
@@ -432,3 +435,47 @@ def perform_destroy(self, instance):
+
+    def delete(self, request, *args, **kwargs):
+        live_event = self.get_object()
+        live_event.status = 'cancelled'
```
`live_event.status = LiveEvent.CANCELED` and define `CANCELED` in `LiveEvent` model class as `CANCELED = 'canceled'.

smirolo commented on pages/mixins.py:
```diff
@@ -144,7 +145,7 @@ def element(self):
 class SequenceMixin(object):
     """
-    Returns an ``User`` from a URL.
+    Returns a ``Sequence`` from a URL.
```
Excellent catch!

smirolo commented on pages/mixins.py:
```diff
@@ -314,12 +317,18 @@ class EnumeratedProgressMixin(SequenceProgressMixin):
 
     rank_url_kwarg = 'rank'
 
+    @property
```
Great refactor!

smirolo commented on pages/models.py:
```diff
@@ -389,13 +389,22 @@ class LiveEvent(models.Model):
+        help_text="Unique integer to denote the index of the LiveEvent")
     location = models.URLField(_("URL to the calendar event"), max_length=2083)
     max_attendees = models.IntegerField(default=0)
+    status = models.CharField(
```
Put the choices as constant defined at the class level. See definition of [FORMAT_CHOICES](https://github.com/djaodjin/djaodjin-pages/blob/096eee41cbbe4e5fd4d2cc0a2c9ab68ce76e57d4/pages/models.py#L172) as example.

smirolo replied:
I meant write code like:

```
   CANCELED = 'Canceled'
   SCHEDULED = 'Scheduled'

   EVENT_STATUS_CHOICES = (
     (SCHEDULED, 'Scheduled'),
     (CANCELED, 'Canceled')
   )
```

such that you write `live_event.status = LiveEvent.CANCELED`

The idea is that using a named "constant", it is a lot easier to maintain the code than if strings are used all over the place. Ideally you would define `SCHEDULED = 0` and `status = models.SmallPositiveInteger` to speed up SQL queries in the database.

smirolo commented on pages/serializers.py:
```diff
@@ -378,8 +378,30 @@ class Meta(EnumeratedElementSerializer.Meta):
+        help_text=_("Username of the attendee"))
+
+    class Meta(EnumeratedProgressSerializer.Meta):
+        fields = ('user', 'content', 'viewing_duration', 'min_viewing_duration')
```
Why is this not `fields = EnumeratedProgressSerializer.fields + ('user', ...`?

Deep-Chill replied:
Because it would show the 'rank' repeatedly for each EnumeratedProgress, even though the rank is provided in the URL and therefore it would be the same for each item. And also moved the user field in the front for easier readability.

smirolo commented on pages/serializers.py:
```diff
@@ -378,8 +378,30 @@ class Meta(EnumeratedElementSerializer.Meta):
+        fields = ('user', 'content', 'viewing_duration', 'min_viewing_duration')
+        read_only_fields = EnumeratedProgressSerializer.Meta.read_only_fields + ('user',)
+
+    def get_user(self, obj):
```
We might want to return a serialized User here at some point so it is easier for the UI to display the full name of the attendee. It has not yet been done in other parts of the API (ex: `Comment` serializer) so this most likely should be note/issue for a future refactoring. Good to keep consistency with existing APIs for now.

smirolo commented on pages/views/sequences.py:
```diff
@@ -71,16 +71,20 @@ def get_context_data(self, **kwargs):
 
     template_name = 'pages/app/sequences/pageelement.html'
-
+    # is_live_event and is_certificate flags not being added
```
What does this comment mean, an issue to be fixed or an implementation note to be aware of?

smirolo reviewed (commented):
There is just really the 'live_event_rank' stored in the `extra` field that require some explanation.

Deep-Chill committed:
Cleans up logic to mark Live Event attendance

Deep-Chill committed:
Minor updates

Deep-Chill committed:
Implements set_extra and other refactors

Uses the set_extra method. It initializes extra as a dictionary if it doesn't exist.

An alternative could be to initialize extra as and empty string if a progress object is created in `Mixins.py` on `line 333`.

smirolo commented on pages/helpers.py:
```diff
@@ -63,6 +63,20 @@ def get_extra(obj, attr_name, default=None):
+        # is to set it to some value.
+        obj.extra = {}
+    obj.extra.update({attr_name: attr_value})
+    obj.extra = json.dumps(obj.extra)
```
We don't want to serialize the extra field here (i.e. `json.dumps`) since the whole idea of `get_extra` is to unserialize it only once. Serialization should happen automatically on `obj.save()`.

smirolo reviewed (approved):
minor optimization regression within the definition of `set_extra`.

Deep-Chill committed:
Removes json.dumps to save obj.extra
